posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py

```diff
@@ -246,28 +262,100 @@ def _update_properties_column_compression_codec(self, query_id, codec):
                 sql=f"ALTER TABLE {settings.CLICKHOUSE_DATABASE}.{EVENTS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}' MODIFY COLUMN {column} VARCHAR Codec({codec})",
             )
 
+    # TODO: Consider making postcheck a native component of the async migrations spec
+    def _postcheck(self, _: str):
+        self._check_person_data()
+        self._check_groups_data()
+
+    def _where_clause(self) -> Tuple[str, Dict[str, Union[str, int]]]:
+        team_id = self.get_parameter("TEAM_ID")
+        team_id_filter = f" AND team_id = %(team_id)s" if team_id else ""
+        where_clause = f"WHERE timestamp > toDateTime(%(timestamp_lower_bound)s) AND timestamp < toDateTime(%(timestamp_upper_bound)s) {team_id_filter}"
+
+        return (
+            where_clause,
+            {
+                "team_id": team_id,
+                "timestamp_lower_bound": self.get_parameter("TIMESTAMP_LOWER_BOUND"),
+                "timestamp_upper_bound": self.get_parameter("TIMESTAMP_UPPER_BOUND"),
+            },
+        )
+
+    def _check_person_data(self, threshold=DEFAULT_ACCEPTED_INCONSISTENT_DATA_RATIO):
+        where_clause, where_clause_params = self._where_clause()
+
+        incomplete_person_data_ratio = sync_execute(
+            f"""
+            SELECT countIf(
+                empty(person_id) OR
+                person_created_at = toDateTime(0) OR
+                person_properties = ''
+            ) / count() FROM events
+            SAMPLE 10000000
+            {where_clause}
+            """,
+            where_clause_params,
+        )[0][0]
+
+        if incomplete_person_data_ratio > threshold:
+            incomplete_events_percentage = incomplete_person_data_ratio * 100
+            raise Exception(
+                f"Backfill did not work succesfully. ~{int(incomplete_events_percentage)}% of events did not get the correct data for persons."
+            )
+
+    def _check_groups_data(self, threshold=DEFAULT_ACCEPTED_INCONSISTENT_DATA_RATIO):
+        where_clause, where_clause_params = self._where_clause()
+
+        # To check if groups data was not backfilled, we check for:
+        # 1. groupX_properties = '' (because the backfill sets group properties to '{}' if the group doesn't exist)
+        # 2. groupX_created_at != created_at column of groups table via dictionary lookup (because we can't just look for groups with toDateTime(0) as that's the default)
+        incomplete_groups_data_ratio = sync_execute(
+            f"""
+            SELECT countIf(
+                group0_properties = '' OR
+                group0_created_at != dictGetDateTime('{settings.CLICKHOUSE_DATABASE}.groups_dict', 'created_at', tuple(team_id, 0, $group_0)) OR
```

Comment:
can't `team_id` be empty (parameter not provided, seeing `team_id_filter` code)?
why is this separate from persons data lookup, I'd suggest instead we do sth like
```
incomplete_person_data_ratio = sync_execute(
            f"""
            SELECT countIf(
                {persons_conditions}
                {groups_conditions}
            ) / count() FROM events
            SAMPLE 10000000
            {where_clause}
            """,
            where_clause_params,
        )[0][0]
        
        
       groups_conditions = "OR ..." or ""
```

Reply:
you're thinking of team_id in different contexts

also this is separate because baking it all into one query is way too heavy and times out. Splitting along the persons/groups line is a split that makes conceptual sense and gives better context to the user